File: contest/olp_khong_chuyen_2019/nhan_ma_tran/main.py
s = input().split()
k = int(s[0])
n = int(s[1])
i = int(s[2])
j = int(s[3])
s = int(s[4])

matrixs = []

for ii in range(0, k):
    matrix = []
    for jj in range(0, n):
        nums = input().split()
        for kk in range(0, n):
            nums[kk] = int(nums[kk])
        matrix.append(nums)
    matrixs.append(matrix)

def mul_matrix(a, b):
    res = [[0] * n] * n
    for jj in range(0, n):
        for kk in range(0, n):
            res[i][jj] += a[i][kk] * b[kk][jj] % s
            res[i][jj] %= s
    return res

res = matrixs[0]
for ii in range(0, n):
    for jj in range(0, n):
        res[ii][jj] %= s

i -= 1
j -= 1
for ii in range(1, k):
    res = mul_matrix(res, matrixs[ii])

print(res[i][j])

File: contest/olp_khong_chuyen_2019/nhan_ma_tran/main.cpp
#include <bits/stdc++.h>
#define ll long long
#define endl '\n'

using namespace std;

int32_t main() {
    ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
    int k, n, i, j;
    ll S;
    cin >> k >> n >> i >> j >> S;
    i--, j--;
    vector<vector<vector<ll>>> matrixs(k, vector<vector<ll>>(n, vector<ll>(n)));
    for(int ii = 0; ii < k; ii++) {
        for(int jj = 0; jj < n; jj++) {
            for(int kk = 0; kk < n; kk++) {
                cin >> matrixs[ii][jj][kk];
                matrixs[ii][jj][kk] %= S;
            }
        }
    }
    auto mul_matrix = [&](vector<vector<ll>> a, vector<vector<ll>> b) -> vector<vector<ll>> {
        vector<vector<ll>> res(n, vector<ll>(n));
        for(int jj = 0; jj < n; jj++) {
            for(int kk = 0; kk < n; kk++) {
                res[i][jj] = (res[i][jj] + a[i][kk] % S * b[kk][jj] % S) % S;
            }
        }
        return res;
    };
    vector<vector<ll>> res = matrixs[0];
    for(int t = 1; t < k; t++) {
        res = mul_matrix(res, matrixs[t]);
    }
    cout << res[i][j];
    return 0;
};